src/connmgr/connection.rs

```diff
@@ -5298,41 +5295,36 @@ where
-
-                                    loop {
-                                        // ABR: select contains read
-                                        let result = select_2(
-                                            add_to_buffer.as_mut(),
-                                            pin!(zsess_in.recv_msg()),
-                                        )
-                                        .await;
-
-                                        match result {
-                                            Select2::R1(ret) => match ret {
-                                                Ok(()) => break,
-                                                Err(CoreHttpError::BufferExceeded) => {
-                                                    return Err(Error::WebSocketRejectionTooLarge(
-                                                        send_buf_size,
-                                                    ));
-                                                }
-                                                Err(e) => return Err(e.into()),
-                                            },
-                                            Select2::R2(ret) => {
-                                                let zreq = ret?;
-
-                                                // ABR: handle_other
-                                                server_handle_other(zreq, zsess_in, zsess_out)
-                                                    .await?;
-                                            }
+                            RecvStatus::Read((), size) => {
+                                body_buf.write_commit(size);
+
+                                if size == 0 {
+                                    return Err(Error::WebSocketRejectionTooLarge(send_buf_size));
+                                }
+                            }
+                            RecvStatus::NeedBytes(()) => {
+                                let mut add_to_buffer = pin!(resp_body.add_to_buffer());
+
+                                loop {
+                                    // ABR: select contains read
+                                    let result =
+                                        select_2(add_to_buffer.as_mut(), pin!(zsess_in.recv_msg()))
+                                            .await;
+
+                                    match result {
+                                        Select2::R1(ret) => {
+                                            ret?;
+                                            break;
+                                        }
+                                        Select2::R2(ret) => {
+                                            let zreq = ret?;
+
+                                            // ABR: handle_other
+                                            server_handle_other(zreq, zsess_in, zsess_out).await?;
                                         }
                                     }
                                 }
```

Comment:
This is similar to earlier code that reads response body bytes from a server into fixed-sized buffer.

This means the only change in behavior is what it does when the output buffer is full. Previously, it would try to read more bytes into the input buffer, either hanging the connection, hanging the thread (busylooping), or eventually returning `WebSocketRejectionTooLarge` if the input buffer filled up also. Now, it immediately returns `WebSocketRejectionTooLarge`. So there should be no performance impact during regular operation, just an earlier error when the output is full and avoiding hangs.